controllers/humiocluster_controller.go

```diff
@@ -2074,10 +2079,170 @@ func (r *HumioClusterReconciler) ensurePodsExist(ctx context.Context, hc *humiov
 }
 
+// Gracefully removes a LogScale pod from the nodepool using the following steps:
+//
+//  1. Matches pod names to node ids
+//  2. Computes the zone from which the pod will be removed base on the current node allocation
+//  3. Iterates through pods and for the first one found in the specified zone, sends an eviction request to the node
+//  4. Checks if the eviction has started (with a timeout of 10 seconds)
+//  5. If the eviction has started, it periodically checks every 60 seconds if the eviction has been completed
+//  6. When the eviction is completed and there is no more data on that node, the node is unregistered from the cluster, and the pod is removed.
+func (r *HumioClusterReconciler) removePodGracefully(ctx context.Context, hc *humiov1alpha1.HumioCluster, podsInNodePool []corev1.Pod, req ctrl.Request) error {
+	// GetCluster gql query returns node ID and Zone
+	clusterConfig, err := helpers.NewCluster(ctx, r, hc.Name, "", hc.Namespace, helpers.UseCertManager(), true, false)
+	if err != nil {
+		return r.logErrorAndReturn(err, fmt.Sprintf("could not create a cluster config for the http client."))
+	}
+	humioHttpClient := r.HumioClient.GetHumioHttpClient(clusterConfig.Config(), req)
+	cluster, err := r.HumioClient.GetCluster(ctx, humioHttpClient, req)
+	if err != nil {
+		return r.logErrorAndReturn(err, "failed to get humio cluster through the GraphQL API")
+	}
+	getCluster := cluster.GetCluster()
+	nodeIdToPodMap := r.matchPodsToHosts(podsInNodePool, getCluster.GetNodes())
+
+	// Check Node Zones and gets the one with the most nodes. In case of a tie, the first zone is used
+	podRemovalZone, err := r.getZoneForPodRemoval(ctx, podsInNodePool)
+	if err != nil {
+		return r.logErrorAndReturn(err, "failed to get pod removal zone")
+	}
+	for _, pod := range podsInNodePool {
+		if pod.Labels[kubernetes.TopologyKubernetesZone] != podRemovalZone {
+			continue
+		}
+		if pod.Spec.NodeName == "" {
+			r.Log.Info(fmt.Sprintf("NodeName is empty for pod %s.", pod.Name))
+			continue
+		}
+		vhost := nodeIdToPodMap[pod.GetName()]
+		err = r.HumioClient.SetIsBeingEvicted(ctx, humioHttpClient, req, vhost, true)
+		if err != nil {
+			return r.logErrorAndReturn(err, fmt.Sprintf("failed to set eviction for vhost %d", vhost))
+		}
+		// wait for eviction status to be updated
+		isEvicted := false
+		for i := 0; i < waitForPodTimeoutSeconds && !isEvicted; i++ {
+			nodesStatus, err := r.getClusterNodesStatus(ctx, humioHttpClient, req)
+			if err != nil {
+				return err
+			}
+
+			for _, node := range nodesStatus {
+				if node.GetId() == vhost && *node.GetIsBeingEvicted() {
+					isEvicted = true
+					break
+				}
+			}
+
+			r.Log.Info(fmt.Sprintf("validating node data eviction is in progress for vhost %d", vhost))
+			time.Sleep(time.Second * 1)
+		}
+
+		if !isEvicted {
+			return r.logErrorAndReturn(err, fmt.Sprintf("failed to set eviction status for vhost %d", vhost))
+		}
+
+		pod.Annotations[kubernetes.PodMarkedForDataEviction] = "true"
```

Comment:
Are we using labels or annotations to manage this? This conflicts with the earlier code that is using label.